src/shared/bus-unit-util.c

```diff
@@ -2937,3 +2938,69 @@ int bus_service_manager_reload(sd_bus *bus) {
 
         return 0;
 }
+
+int unit_freezer_recover(const char *name, UnitFreezer *ret) {
+        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_free_ char *namedup = NULL;
+        int r;
+
+        assert(name);
+        assert(ret);
+
+        namedup = strdup(name);
+        if (!namedup)
+                return log_oom_debug();
+
+        r = bus_connect_system_systemd(&bus);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open connection to systemd: %m");
+
+        (void) sd_bus_set_method_call_timeout(bus, FREEZE_TIMEOUT);
+
+        *ret = (UnitFreezer) {
+                .name = TAKE_PTR(namedup),
+                .bus = TAKE_PTR(bus),
+        };
+        return 0;
+}
+
+void unit_freezer_clear(UnitFreezer *f) {
```

Comment:
we usually call "shallow" destructors like this xyz_done()

Reply:
So it's technically a shallow destructor, yes, but API-wise it's not really supposed to be used like one

It's really supposed to be used like:

```c
_cleanup_(unit_freezer_thaw) UnitFreezer freezer = UNIT_FREEZER_NULL;

unit_freezer_freeze("my-unit.service", &freezer);

r = do_something()
if (r < 0)
    return r; /* my-unit.service is thawed */

r = do_something_else()
if (r < 0)
    return r; /* my-unit.service is thawed */

unit_freezer_clear(&freezer);
/* my-unit.service will remain frozen */
```

So perhaps `unit_freezer_cancel` is a better name for how it's used?